engine/src/universe_generic has a compile warning

[ 21%] Building CXX object CMakeFiles/engine_com.dir/src/universe_generic.cpp.o
/home/evert/vsmod/git/Vega-Strike-Engine-Source/engine/src/universe_generic.cpp: In member function 'void Universe::clearAllSystems()':
/home/evert/vsmod/git/Vega-Strike-Engine-Source/engine/src/universe_generic.cpp:36:9: warning: deleting object of polymorphic class type 'StarSystem' which has non-virtual destructor might cause undefined behavior [-Wdelete-non-virtual-dtor]
   36 |         delete star_system.back();
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~
[ 22%] Building CXX object CMakeFiles/engine_com.dir/src/universe_util_generic.cpp.o

--- a/engine/CMakeLists.txt
+++ b/engine/CMakeLists.txt
@@ -531,9 +531,9 @@ IF (NOT BUILD_OPT)
     SET(BUILD_OPT -O2 CACHE STRING "-O0, -O1, -O2, -O3, -Os, -Ofast" FORCE )
 ENDIF (NOT BUILD_OPT)
 
-IF (CMAKE_COMPILER_IS_GNUCXX)
-    ADD_DEFINITIONS(-std=gnu++03)
-ENDIF ()
+# IF (CMAKE_COMPILER_IS_GNUCXX)
+#     ADD_DEFINITIONS(-std=gnu++03)
+# ENDIF ()
 
 
 # Provide boolean options for enabling various cpu optimizations.

--- a/engine/src/cmd/basecomputer.cpp
+++ b/engine/src/cmd/basecomputer.cpp
@@ -396,7 +396,7 @@ extern float RepairPrice( float operational, float price );
 
 static float basicRepairPrice( void )
 {
-    static const float price = XMLSupport::parse_float( vs_config->getVariable( "physics", "repair_price", "5000" ) );
+    static float price = XMLSupport::parse_float( vs_config->getVariable( "physics", "repair_price", "5000" ) );
     return price*g_game.difficulty;
 }
 
@@ -2807,7 +2807,7 @@ void BaseComputer::loadNewsControls( void )
     picker->clear();
 
     //Load the picker.
-    static const bool newsFromCargolist =
+    static bool newsFromCargolist =
         XMLSupport::parse_bool( vs_config->getVariable( "cargo", "news_from_cargolist", "false" ) );
     if (newsFromCargolist ) {
         gameMessage last;
@@ -3141,7 +3141,7 @@ void BaseComputer::loadSellUpgradeControls( void )
     loadMasterList( partListUnit, weapfiltervec, std::vector< std::string > (), false, tlist );
     ClearDowngradeMap();
     playerUnit->FilterDowngradeList( tlist.masterList );
-    static const bool clearDowngrades =
+    static bool clearDowngrades =
         XMLSupport::parse_bool( vs_config->getVariable( "physics", "only_show_best_downgrade", "true" ) );
     if (clearDowngrades) {
         std::set< std::string >downgradeMap = GetListOfDowngrades();
@@ -3863,11 +3863,11 @@ Cargo CreateCargoForOwnerStarship( const Cockpit *cockpit, const Unit *base, int
     bool needsJumpTransport = (locationSystemName != destinationSystemName);
     bool needsInsysTransport = (locationBaseName != destinationBaseName);
 
-    static const float shipping_price_base =
+    static float shipping_price_base =
         XMLSupport::parse_float( vs_config->getVariable( "physics", "shipping_price_base", "0" ) );
-    static const float shipping_price_insys =
+    static float shipping_price_insys =
         XMLSupport::parse_float( vs_config->getVariable( "physics", "shipping_price_insys", "1000" ) );
-    static const float shipping_price_perjump =
+    static float shipping_price_perjump =
         XMLSupport::parse_float( vs_config->getVariable( "physics", "shipping_price_perjump", "25000" ) );
 
     cargo.price = shipping_price_base;
@@ -4281,7 +4281,7 @@ bool sellShip( Unit *baseUnit, Unit *playerUnit, std::string shipname, BaseCompu
 
                 float xtra = 0;
                 if ( cockpit->GetUnitSystemName(i) == _Universe->activeStarSystem()->getFileName() ) {
-                    static const float shipping_price =
+                    static float shipping_price =
                         XMLSupport::parse_float( vs_config->getVariable( "physics", "sellback_shipping_price", "6000" ) );
                     xtra += shipping_price;
                 }

--- a/engine/src/cmd/collision.cpp
+++ b/engine/src/cmd/collision.cpp
@@ -11,42 +11,42 @@
 
 // TODO: convert all float to double and all Vector to QVector.
 
-static const float kilojoules_per_damage = GameConfig::GetVariable( "physics", "kilojoules_per_unit_damage", 5400 );
-static const float collision_scale_factor =
+static float kilojoules_per_damage = GameConfig::GetVariable( "physics", "kilojoules_per_unit_damage", 5400 );
+static float collision_scale_factor =
         GameConfig::GetVariable( "physics", "collision_damage_scale", 1.0f );
-static const float inelastic_scale = GameConfig::GetVariable( "physics", "inelastic_scale", 0.8f);
-static const float min_time =
+static float inelastic_scale = GameConfig::GetVariable( "physics", "inelastic_scale", 0.8f);
+static float min_time =
         GameConfig::GetVariable( "physics", "minimum_time_between_recorded_player_collisions", 0.1f);
 static const float minimum_mass = 1e-6f;
 
 //Collision force caps primarily for AI-AI collisions. Once the AIs get a real collision avoidance system, we can
 // turn damage for AI-AI collisions back on, and then we can remove these caps.
 //value, in seconds of desired maximum recovery time
-static const float max_torque_multiplier =
+static float max_torque_multiplier =
         GameConfig::GetVariable( "physics", "maxCollisionTorqueMultiplier", 0.67f);
 //value, in seconds of desired maximum recovery time
-static const float max_force_multiplier  =
+static float max_force_multiplier  =
         GameConfig::GetVariable( "physics", "maxCollisionForceMultiplier", 5);
 
-static const int upgrade_faction =
+static int upgrade_faction =
         GameConfig::GetVariable( "physics", "cargo_deals_collide_damage",
                                                         false) ? -1 : FactionUtil::GetUpgradeFaction();
 
-static const float collision_hack_distance =
+static float collision_hack_distance =
         GameConfig::GetVariable( "physics", "collision_avoidance_hack_distance", 10000);
-static const float front_collision_hack_distance =
+static float front_collision_hack_distance =
         GameConfig::GetVariable( "physics", "front_collision_avoidance_hack_distance", 200000);
 
 static float front_collision_hack_angle = cos( 3.1415926536f * GameConfig::GetVariable( "physics", "front_collision_avoidance_hack_angle", 40)/180.0f );
 
-static const bool collision_damage_to_ai = GameConfig::GetVariable( "physics", "collisionDamageToAI", false);
+static bool collision_damage_to_ai = GameConfig::GetVariable( "physics", "collisionDamageToAI", false);
 
-static const bool crash_dock_unit = GameConfig::GetVariable( "physics", "unit_collision_docks", false);
+static bool crash_dock_unit = GameConfig::GetVariable( "physics", "unit_collision_docks", false);
 
-static const bool crash_dock_hangar = GameConfig::GetVariable( "physics", "only_hangar_collision_docks", false);
+static bool crash_dock_hangar = GameConfig::GetVariable( "physics", "only_hangar_collision_docks", false);
 
 // Disabled bouncing missile option. Missiles always explode when colliding with something.
-//static const bool does_missile_bounce = GameConfig::GetVariable( "physics", "missile_bounce", false);
+//static bool does_missile_bounce = GameConfig::GetVariable( "physics", "missile_bounce", false);
 
 
 

--- a/engine/src/cmd/damageable.h
+++ b/engine/src/cmd/damageable.h
@@ -20,6 +20,14 @@ public:
 public:
   Damageable();
 
+protected:
+  virtual ~Damageable() = default;
+  // forbidden
+  Damageable( const Damageable& ) = delete;
+  // forbidden
+  Damageable& operator=( const Damageable& ) = delete;
+
+public:
   float GetHull() const { return hull;}
 
   void ArmorData( float armor[8] ) const;

--- a/engine/src/cmd/drawable.h
+++ b/engine/src/cmd/drawable.h
@@ -53,8 +53,14 @@ public:
 
     void clear();
 
-    ~Drawable() { clear(); }
+protected:
+    virtual ~Drawable() { clear(); }
+    // forbidden
+    Drawable( const Drawable& ) = delete;
+    // forbidden
+    Drawable& operator=( const Drawable& ) = delete;
 
+public:
     string getAnimationName(unsigned int animationNumber) const;
 
     unsigned int getAnimationNumber(const char *name) const;

--- a/engine/src/cmd/movable.cpp
+++ b/engine/src/cmd/movable.cpp
@@ -266,14 +266,16 @@ Vector Movable::ResolveForces( const Transformation &trans, const Matrix &transm
         AngularVelocity = AngularVelocity.Normalize()*caprate;
     //acceleration
     Vector temp2 = (NetLocalForce.i*p+NetLocalForce.j*q+NetLocalForce.k*r);
-    if ( !( FINITE( NetForce.i ) && FINITE( NetForce.j ) && FINITE( NetForce.k ) ) )
-        std::cout<<"NetForce skrewed";
+    if ( !( FINITE( NetForce.i ) && FINITE( NetForce.j ) && FINITE( NetForce.k ) ) ) {
+        BOOST_LOG_TRIVIAL(info) << "NetForce skrewed";
+    }
     if (NetForce.i || NetForce.j || NetForce.k)
         temp2 += InvTransformNormal( transmat, NetForce );
     temp2 = temp2/GetMass();
     temp  = temp2*simulation_atom_var;
-    if ( !( FINITE( temp2.i ) && FINITE( temp2.j ) && FINITE( temp2.k ) ) )
-        std::cout<<"NetForce transform skrewed";
+    if ( !( FINITE( temp2.i ) && FINITE( temp2.j ) && FINITE( temp2.k ) ) ) {
+        BOOST_LOG_TRIVIAL(info) << "NetForce transform skrewed";
+    }
     float oldmagsquared = Velocity.MagnitudeSquared();
         Velocity += temp;
     //}

--- a/engine/src/cmd/movable.cpp
+++ b/engine/src/cmd/movable.cpp
@@ -9,33 +9,33 @@
 
 #include <iostream>
 
-static const float VELOCITY_MAX = GameConfig::GetVariable( "physics", "velocity_max", 10000);
+static float VELOCITY_MAX = GameConfig::GetVariable( "physics", "velocity_max", 10000);
 
 //for the heck of it.
-static const float humanwarprampuptime = GameConfig::GetVariable( "physics", "warprampuptime", 5);
+static float humanwarprampuptime = GameConfig::GetVariable( "physics", "warprampuptime", 5);
 
 //for the heck of it.
-static const float compwarprampuptime = GameConfig::GetVariable( "physics", "computerwarprampuptime", 10);
+static float compwarprampuptime = GameConfig::GetVariable( "physics", "computerwarprampuptime", 10);
 
-static const float warprampdowntime = GameConfig::GetVariable( "physics", "warprampdowntime", 0.5f);
-static const float WARPMEMORYEFFECT = GameConfig::GetVariable( "physics", "WarpMemoryEffect", 0.9f);
-static const float maxplayerrotationrate = GameConfig::GetVariable( "physics", "maxplayerrot", 24);
-static const float maxnonplayerrotationrate = GameConfig::GetVariable( "physics", "maxNPCrot", 360);
-static const float warpstretchcutoff = GameConfig::GetVariable( "graphics", "warp_stretch_cutoff", 500000) * GameConfig::GetVariable( "physics", "game_speed", 1);
-static const float warpstretchoutcutoff = GameConfig::GetVariable( "graphics", "warp_stretch_decel_cutoff", 500000) * GameConfig::GetVariable( "physics", "game_speed", 1);
-static const float sec = GameConfig::GetVariable( "graphics", "insys_jump_ani_second_ahead", 4) / (GameConfig::GetVariable( "physics", "game_speed", 1) * GameConfig::GetVariable( "physics", "game_accel", 1));
-static const float endsec = GameConfig::GetVariable( "graphics", "insys_jump_ani_second_ahead_end",0.03f) /( GameConfig::GetVariable( "physics", "game_speed",1.0f) *GameConfig::GetVariable( "physics", "game_accel", 1.0f) );
+static float warprampdowntime = GameConfig::GetVariable( "physics", "warprampdowntime", 0.5f);
+static float WARPMEMORYEFFECT = GameConfig::GetVariable( "physics", "WarpMemoryEffect", 0.9f);
+static float maxplayerrotationrate = GameConfig::GetVariable( "physics", "maxplayerrot", 24);
+static float maxnonplayerrotationrate = GameConfig::GetVariable( "physics", "maxNPCrot", 360);
+static float warpstretchcutoff = GameConfig::GetVariable( "graphics", "warp_stretch_cutoff", 500000) * GameConfig::GetVariable( "physics", "game_speed", 1);
+static float warpstretchoutcutoff = GameConfig::GetVariable( "graphics", "warp_stretch_decel_cutoff", 500000) * GameConfig::GetVariable( "physics", "game_speed", 1);
+static float sec = GameConfig::GetVariable( "graphics", "insys_jump_ani_second_ahead", 4) / (GameConfig::GetVariable( "physics", "game_speed", 1) * GameConfig::GetVariable( "physics", "game_accel", 1));
+static float endsec = GameConfig::GetVariable( "graphics", "insys_jump_ani_second_ahead_end",0.03f) /( GameConfig::GetVariable( "physics", "game_speed",1.0f) *GameConfig::GetVariable( "physics", "game_accel", 1.0f) );
 
 
 //Pi^2
 
-static const float  warpMultiplierMin = GameConfig::GetVariable( "physics", "warpMultiplierMin", 9.86960440109f);
+static float  warpMultiplierMin = GameConfig::GetVariable( "physics", "warpMultiplierMin", 9.86960440109f);
 
 //C
-static const float  warpMultiplierMax = GameConfig::GetVariable( "physics", "warpMultiplierMax", 300000000 );
+static float  warpMultiplierMax = GameConfig::GetVariable( "physics", "warpMultiplierMax", 300000000 );
 
 //Pi^2 * C
-static const float warpMaxEfVel = GameConfig::GetVariable( "physics", "warpMaxEfVel", 2960881320.0f);
+static float warpMaxEfVel = GameConfig::GetVariable( "physics", "warpMaxEfVel", 2960881320.0f);
 
 
 
@@ -315,8 +315,8 @@ Vector Movable::ResolveForces( const Transformation &trans, const Matrix &transm
         UniverseUtil::playAnimationGrow( insys_jump_ani, realPosition().Cast()+Velocity*tmpsec+v*radial_size, radial_size*8, 1 );
     }
 
-    static const float air_res_coef = XMLSupport::parse_float( active_missions[0]->getVariable( "air_resistance", "0" ) );
-    static const float lateral_air_res_coef = XMLSupport::parse_float( active_missions[0]->getVariable( "lateral_air_resistance", "0"));
+    static float air_res_coef = XMLSupport::parse_float( active_missions[0]->getVariable( "air_resistance", "0" ) );
+    static float lateral_air_res_coef = XMLSupport::parse_float( active_missions[0]->getVariable( "lateral_air_resistance", "0"));
 
     if (air_res_coef || lateral_air_res_coef) {
         float  velmag = Velocity.Magnitude();

--- a/engine/src/cmd/movable.cpp
+++ b/engine/src/cmd/movable.cpp
@@ -8,10 +8,60 @@
 #include "game_config.h"
 
 #include <iostream>
+#include <string>
 
 
 
-Movable::Movable() : cumulative_transformation_matrix( identity_matrix ) {}
+bool Movable::configLoaded = false;
+float Movable::VELOCITY_MAX = 0.0f;
+//for the heck of it.
+float Movable::humanwarprampuptime = 0.0f;
+//for the heck of it.
+float Movable::compwarprampuptime = 0.0f;
+float Movable::warprampdowntime = 0.0f;
+float Movable::WARPMEMORYEFFECT = 0.0f;
+float Movable::maxplayerrotationrate = 0.0f;
+float Movable::maxnonplayerrotationrate = 0.0f;
+float Movable::warpstretchcutoff = 0.0f;
+float Movable::warpstretchoutcutoff = 0.0f;
+float Movable::sec = 0.0f;
+float Movable::endsec = 0.0f;
+float Movable::warpMultiplierMin = 0.0f;
+float Movable::warpMultiplierMax = 0.0f;
+float Movable::warpMaxEfVel = 0.0f;
+float Movable::cutsqr = 0.0f;
+float Movable::outcutsqr = 0.0f;
+std::string Movable::insys_jump_ani = "";
+float Movable::air_res_coef = 0.0f;
+float Movable::lateral_air_res_coef = 0.0f;
+
+Movable::Movable() : cumulative_transformation_matrix( identity_matrix ) {
+    if (!configLoaded) {
+        VELOCITY_MAX = GameConfig::GetVariable( "physics", "velocity_max", 10000);
+        humanwarprampuptime = GameConfig::GetVariable( "physics", "warprampuptime", 5);
+        compwarprampuptime = GameConfig::GetVariable( "physics", "computerwarprampuptime", 10);
+        warprampdowntime = GameConfig::GetVariable( "physics", "warprampdowntime", 0.5f);
+        WARPMEMORYEFFECT = GameConfig::GetVariable( "physics", "WarpMemoryEffect", 0.9f);
+        maxplayerrotationrate = GameConfig::GetVariable( "physics", "maxplayerrot", 24);
+        maxnonplayerrotationrate = GameConfig::GetVariable( "physics", "maxNPCrot", 360);
+        warpstretchcutoff = GameConfig::GetVariable( "graphics", "warp_stretch_cutoff", 500000) * GameConfig::GetVariable( "physics", "game_speed", 1);
+        warpstretchoutcutoff = GameConfig::GetVariable( "graphics", "warp_stretch_decel_cutoff", 500000) * GameConfig::GetVariable( "physics", "game_speed", 1);
+        sec = GameConfig::GetVariable( "graphics", "insys_jump_ani_second_ahead", 4) / (GameConfig::GetVariable( "physics", "game_speed", 1) * GameConfig::GetVariable( "physics", "game_accel", 1));
+        endsec = GameConfig::GetVariable( "graphics", "insys_jump_ani_second_ahead_end",0.03f) /( GameConfig::GetVariable( "physics", "game_speed",1.0f) *GameConfig::GetVariable( "physics", "game_accel", 1.0f) );
+        //Pi^2
+        warpMultiplierMin = GameConfig::GetVariable( "physics", "warpMultiplierMin", 9.86960440109f);
+        //C
+        warpMultiplierMax = GameConfig::GetVariable( "physics", "warpMultiplierMax", 300000000 );
+        //Pi^2 * C
+        warpMaxEfVel = GameConfig::GetVariable( "physics", "warpMaxEfVel", 2960881320.0f);
+        cutsqr    = warpstretchcutoff * warpstretchcutoff;
+        outcutsqr = warpstretchoutcutoff * warpstretchoutcutoff;
+        insys_jump_ani = vs_config->getVariable( "graphics", "insys_jump_animation", "warp.ani" );
+        air_res_coef = XMLSupport::parse_float( active_missions[0]->getVariable( "air_resistance", "0" ) );
+        lateral_air_res_coef = XMLSupport::parse_float( active_missions[0]->getVariable( "lateral_air_resistance", "0"));
+        configLoaded = true;
+    }
+}
 
 
 
@@ -89,11 +139,6 @@ void Movable::UpdatePhysics( const Transformation &trans,
                           UnitCollection *uc,
                           Unit *superunit )
 {
-    static float VELOCITY_MAX = GameConfig::GetVariable( "physics", "velocity_max", 10000);
-
-
-
-
     //Save information about when this happened
     unsigned int cur_sim_frame = _Universe->activeStarSystem()->getCurrentSimFrame();
     //Well, wasn't skipped actually, but...
@@ -134,13 +179,6 @@ void Movable::UpdatePhysics( const Transformation &trans,
 
 void Movable::AddVelocity( float difficulty )
 {
-    //for the heck of it.
-    static float humanwarprampuptime = GameConfig::GetVariable( "physics", "warprampuptime", 5);
-    //for the heck of it.
-    static float compwarprampuptime = GameConfig::GetVariable( "physics", "computerwarprampuptime", 10);
-    static float warprampdowntime = GameConfig::GetVariable( "physics", "warprampdowntime", 0.5f);
-    static float WARPMEMORYEFFECT = GameConfig::GetVariable( "physics", "WarpMemoryEffect", 0.9f);
-
     float  lastWarpField = graphicOptions.WarpFieldStrength;
 
     bool   playa = isPlayerShip();
@@ -232,13 +270,6 @@ void Movable::Rotate( const Vector &axis )
 
 Vector Movable::ResolveForces( const Transformation &trans, const Matrix &transmat )
 {
-    static float maxplayerrotationrate = GameConfig::GetVariable( "physics", "maxplayerrot", 24);
-    static float maxnonplayerrotationrate = GameConfig::GetVariable( "physics", "maxNPCrot", 360);
-    static float warpstretchcutoff = GameConfig::GetVariable( "graphics", "warp_stretch_cutoff", 500000) * GameConfig::GetVariable( "physics", "game_speed", 1);
-    static float warpstretchoutcutoff = GameConfig::GetVariable( "graphics", "warp_stretch_decel_cutoff", 500000) * GameConfig::GetVariable( "physics", "game_speed", 1);
-    static float sec = GameConfig::GetVariable( "graphics", "insys_jump_ani_second_ahead", 4) / (GameConfig::GetVariable( "physics", "game_speed", 1) * GameConfig::GetVariable( "physics", "game_accel", 1));
-    static float endsec = GameConfig::GetVariable( "graphics", "insys_jump_ani_second_ahead_end",0.03f) /( GameConfig::GetVariable( "physics", "game_speed",1.0f) *GameConfig::GetVariable( "physics", "game_accel", 1.0f) );
-
     //First, save theoretical instantaneous acceleration (not time-quantized) for GetAcceleration()
     SavedAccel = GetNetAcceleration();
     SavedAngAccel = GetNetAngularAcceleration();
@@ -282,14 +313,11 @@ Vector Movable::ResolveForces( const Transformation &trans, const Matrix &transm
 
     float newmagsquared = Velocity.MagnitudeSquared();
 
-    static float cutsqr    = warpstretchcutoff*warpstretchcutoff;
-    static float outcutsqr = warpstretchoutcutoff*warpstretchoutcutoff;
     bool oldbig    = oldmagsquared > cutsqr;
     bool newbig    = newmagsquared > cutsqr;
     bool oldoutbig = oldmagsquared > outcutsqr;
     bool newoutbig = newmagsquared > outcutsqr;
     if ( (newbig && !oldbig) || (oldoutbig && !newoutbig) ) {
-        static string insys_jump_ani = vs_config->getVariable( "graphics", "insys_jump_animation", "warp.ani" );
         static bool   docache = true;
         if (docache) {
             UniverseUtil::cacheAnimation( insys_jump_ani );
@@ -303,9 +331,6 @@ Vector Movable::ResolveForces( const Transformation &trans, const Matrix &transm
         float tmpsec = oldbig ? endsec : sec;
         UniverseUtil::playAnimationGrow( insys_jump_ani, realPosition().Cast()+Velocity*tmpsec+v*radial_size, radial_size*8, 1 );
     }
-
-    static float air_res_coef = XMLSupport::parse_float( active_missions[0]->getVariable( "air_resistance", "0" ) );
-    static float lateral_air_res_coef = XMLSupport::parse_float( active_missions[0]->getVariable( "lateral_air_resistance", "0"));
 
     if (air_res_coef || lateral_air_res_coef) {
         float  velmag = Velocity.Magnitude();
@@ -397,15 +422,6 @@ Vector Movable::ToWorldCoordinates( const Vector &v ) const
 
 float Movable::GetMaxWarpFieldStrength( float rampmult ) const
 {
-    //Pi^2
-    static float warpMultiplierMin = GameConfig::GetVariable( "physics", "warpMultiplierMin", 9.86960440109f);
-
-    //C
-    static float warpMultiplierMax = GameConfig::GetVariable( "physics", "warpMultiplierMax", 300000000 );
-
-    //Pi^2 * C
-    static float warpMaxEfVel = GameConfig::GetVariable( "physics", "warpMaxEfVel", 2960881320.0f);
-
     Vector v = GetWarpRefVelocity();
 
 

--- a/engine/src/cmd/movable.cpp
+++ b/engine/src/cmd/movable.cpp
@@ -9,34 +9,6 @@
 
 #include <iostream>
 
-static float VELOCITY_MAX = GameConfig::GetVariable( "physics", "velocity_max", 10000);
-
-//for the heck of it.
-static float humanwarprampuptime = GameConfig::GetVariable( "physics", "warprampuptime", 5);
-
-//for the heck of it.
-static float compwarprampuptime = GameConfig::GetVariable( "physics", "computerwarprampuptime", 10);
-
-static float warprampdowntime = GameConfig::GetVariable( "physics", "warprampdowntime", 0.5f);
-static float WARPMEMORYEFFECT = GameConfig::GetVariable( "physics", "WarpMemoryEffect", 0.9f);
-static float maxplayerrotationrate = GameConfig::GetVariable( "physics", "maxplayerrot", 24);
-static float maxnonplayerrotationrate = GameConfig::GetVariable( "physics", "maxNPCrot", 360);
-static float warpstretchcutoff = GameConfig::GetVariable( "graphics", "warp_stretch_cutoff", 500000) * GameConfig::GetVariable( "physics", "game_speed", 1);
-static float warpstretchoutcutoff = GameConfig::GetVariable( "graphics", "warp_stretch_decel_cutoff", 500000) * GameConfig::GetVariable( "physics", "game_speed", 1);
-static float sec = GameConfig::GetVariable( "graphics", "insys_jump_ani_second_ahead", 4) / (GameConfig::GetVariable( "physics", "game_speed", 1) * GameConfig::GetVariable( "physics", "game_accel", 1));
-static float endsec = GameConfig::GetVariable( "graphics", "insys_jump_ani_second_ahead_end",0.03f) /( GameConfig::GetVariable( "physics", "game_speed",1.0f) *GameConfig::GetVariable( "physics", "game_accel", 1.0f) );
-
-
-//Pi^2
-
-static float  warpMultiplierMin = GameConfig::GetVariable( "physics", "warpMultiplierMin", 9.86960440109f);
-
-//C
-static float  warpMultiplierMax = GameConfig::GetVariable( "physics", "warpMultiplierMax", 300000000 );
-
-//Pi^2 * C
-static float warpMaxEfVel = GameConfig::GetVariable( "physics", "warpMaxEfVel", 2960881320.0f);
-
 
 
 Movable::Movable() : cumulative_transformation_matrix( identity_matrix ) {}
@@ -117,6 +89,8 @@ void Movable::UpdatePhysics( const Transformation &trans,
                           UnitCollection *uc,
                           Unit *superunit )
 {
+    static float VELOCITY_MAX = GameConfig::GetVariable( "physics", "velocity_max", 10000);
+
 
 
 
@@ -160,6 +134,12 @@ void Movable::UpdatePhysics( const Transformation &trans,
 
 void Movable::AddVelocity( float difficulty )
 {
+    //for the heck of it.
+    static float humanwarprampuptime = GameConfig::GetVariable( "physics", "warprampuptime", 5);
+    //for the heck of it.
+    static float compwarprampuptime = GameConfig::GetVariable( "physics", "computerwarprampuptime", 10);
+    static float warprampdowntime = GameConfig::GetVariable( "physics", "warprampdowntime", 0.5f);
+    static float WARPMEMORYEFFECT = GameConfig::GetVariable( "physics", "WarpMemoryEffect", 0.9f);
 
     float  lastWarpField = graphicOptions.WarpFieldStrength;
 
@@ -252,6 +232,13 @@ void Movable::Rotate( const Vector &axis )
 
 Vector Movable::ResolveForces( const Transformation &trans, const Matrix &transmat )
 {
+    static float maxplayerrotationrate = GameConfig::GetVariable( "physics", "maxplayerrot", 24);
+    static float maxnonplayerrotationrate = GameConfig::GetVariable( "physics", "maxNPCrot", 360);
+    static float warpstretchcutoff = GameConfig::GetVariable( "graphics", "warp_stretch_cutoff", 500000) * GameConfig::GetVariable( "physics", "game_speed", 1);
+    static float warpstretchoutcutoff = GameConfig::GetVariable( "graphics", "warp_stretch_decel_cutoff", 500000) * GameConfig::GetVariable( "physics", "game_speed", 1);
+    static float sec = GameConfig::GetVariable( "graphics", "insys_jump_ani_second_ahead", 4) / (GameConfig::GetVariable( "physics", "game_speed", 1) * GameConfig::GetVariable( "physics", "game_accel", 1));
+    static float endsec = GameConfig::GetVariable( "graphics", "insys_jump_ani_second_ahead_end",0.03f) /( GameConfig::GetVariable( "physics", "game_speed",1.0f) *GameConfig::GetVariable( "physics", "game_accel", 1.0f) );
+
     //First, save theoretical instantaneous acceleration (not time-quantized) for GetAcceleration()
     SavedAccel = GetNetAcceleration();
     SavedAngAccel = GetNetAngularAcceleration();
@@ -408,6 +395,15 @@ Vector Movable::ToWorldCoordinates( const Vector &v ) const
 
 float Movable::GetMaxWarpFieldStrength( float rampmult ) const
 {
+    //Pi^2
+    static float warpMultiplierMin = GameConfig::GetVariable( "physics", "warpMultiplierMin", 9.86960440109f);
+
+    //C
+    static float warpMultiplierMax = GameConfig::GetVariable( "physics", "warpMultiplierMax", 300000000 );
+
+    //Pi^2 * C
+    static float warpMaxEfVel = GameConfig::GetVariable( "physics", "warpMaxEfVel", 2960881320.0f);
+
     Vector v = GetWarpRefVelocity();
 
 

--- a/engine/src/cmd/movable.cpp
+++ b/engine/src/cmd/movable.cpp
@@ -57,8 +57,6 @@ Movable::Movable() : cumulative_transformation_matrix( identity_matrix ) {
         cutsqr    = warpstretchcutoff * warpstretchcutoff;
         outcutsqr = warpstretchoutcutoff * warpstretchoutcutoff;
         insys_jump_ani = vs_config->getVariable( "graphics", "insys_jump_animation", "warp.ani" );
-        air_res_coef = XMLSupport::parse_float( active_missions[0]->getVariable( "air_resistance", "0" ) );
-        lateral_air_res_coef = XMLSupport::parse_float( active_missions[0]->getVariable( "lateral_air_resistance", "0"));
         configLoaded = true;
     }
 }
@@ -331,6 +329,10 @@ Vector Movable::ResolveForces( const Transformation &trans, const Matrix &transm
         float tmpsec = oldbig ? endsec : sec;
         UniverseUtil::playAnimationGrow( insys_jump_ani, realPosition().Cast()+Velocity*tmpsec+v*radial_size, radial_size*8, 1 );
     }
+
+    // stephengtuggy 2020-10-17: These need to be initialized here, because they depend on having an active mission.
+    air_res_coef = XMLSupport::parse_float( active_missions[0]->getVariable( "air_resistance", "0" ) );
+    lateral_air_res_coef = XMLSupport::parse_float( active_missions[0]->getVariable( "lateral_air_resistance", "0"));
 
     if (air_res_coef || lateral_air_res_coef) {
         float  velmag = Velocity.Magnitude();

--- a/engine/src/cmd/movable.h
+++ b/engine/src/cmd/movable.h
@@ -89,6 +89,28 @@ protected:
     float  Momentofinertia;
     Vector SavedAccel;
     Vector SavedAngAccel;
+    static bool configLoaded;
+    static float VELOCITY_MAX;
+    //for the heck of it.
+    static float humanwarprampuptime;
+    //for the heck of it.
+    static float compwarprampuptime;
+    static float warprampdowntime;
+    static float WARPMEMORYEFFECT;
+    static float maxplayerrotationrate;
+    static float maxnonplayerrotationrate;
+    static float warpstretchcutoff;
+    static float warpstretchoutcutoff;
+    static float sec;
+    static float endsec;
+    static float warpMultiplierMin;
+    static float warpMultiplierMax;
+    static float warpMaxEfVel;
+    static float cutsqr;
+    static float outcutsqr;
+    static std::string insys_jump_ani;
+    static float air_res_coef;
+    static float lateral_air_res_coef;
 
 
 // Methods
@@ -96,6 +118,14 @@ protected:
 public:
     Movable();
 
+protected:
+    // forbidden
+    Movable( const Movable& ) = delete;
+    // forbidden
+    Movable& operator= ( const Movable& ) = delete;
+    virtual ~Movable() = default;
+
+public:
     void AddVelocity( float difficulty );
 //Resolves forces of given unit on a physics frame
     virtual Vector ResolveForces( const Transformation&, const Matrix& );

--- a/engine/src/cmd/unit_generic.cpp
+++ b/engine/src/cmd/unit_generic.cpp
@@ -664,7 +664,7 @@ std::string lookupMountSize( int s )
  **** UNIT STUFF
  **********************************************************************************
  */
-Unit::Unit( int /*dummy*/ )
+Unit::Unit( int /*dummy*/ ) : Drawable(), Damageable(), Movable()
 {
     ZeroAll();
     pImage  = (new UnitImages< void >);
@@ -675,7 +675,7 @@ Unit::Unit( int /*dummy*/ )
     Init();
 }
 
-Unit::Unit() //: cumulative_transformation_matrix( identity_matrix )
+Unit::Unit() : Drawable(), Damageable(), Movable() //: cumulative_transformation_matrix( identity_matrix )
 {
     ZeroAll();
     pImage  = (new UnitImages< void >);
@@ -686,7 +686,7 @@ Unit::Unit() //: cumulative_transformation_matrix( identity_matrix )
     Init();
 }
 
-Unit::Unit( std::vector< Mesh* > &meshes, bool SubU, int fact ) //: cumulative_transformation_matrix( identity_matrix )
+Unit::Unit( std::vector< Mesh* > &meshes, bool SubU, int fact ) : Drawable(), Damageable(), Movable() //: cumulative_transformation_matrix( identity_matrix )
 {
     ZeroAll();
     pImage  = (new UnitImages< void >);
@@ -713,7 +713,7 @@ Unit::Unit( const char *filename,
             std::string unitModifications,
             Flightgroup *flightgrp,
             int fg_subnumber,
-            string *netxml ) //: cumulative_transformation_matrix( identity_matrix )
+            string *netxml ) : Drawable(), Damageable(), Movable() //: cumulative_transformation_matrix( identity_matrix )
 {
     ZeroAll();
     pImage  = (new UnitImages< void >);

--- a/engine/src/cmd/unit_generic.h
+++ b/engine/src/cmd/unit_generic.h
@@ -247,12 +247,15 @@ public:
 
 protected:
 //forbidden
-    Unit( const Unit& );
+    Unit( const Unit& ) = delete;
 
 //forbidden
-    Unit& operator=( const Unit& );
+    Unit& operator=( const Unit& ) = delete;
 
-public: Unit();
+    virtual ~Unit();
+
+public: 
+    Unit();
 
 /** Default constructor. This is just to figure out where default
  *  constructors are used. The useless argument will be removed
@@ -286,7 +289,6 @@ public:
     friend class UnitFactory;
 //table can be NULL, but setting it appropriately may increase performance
     void LoadRow( class CSVRow&row, std::string unitMod, std::string*netxml = NULL );
-    virtual ~Unit();
 
     // TODO: implement enum class as type safe bitmask...
     // http://blog.bitwigglers.org/using-enum-classes-as-type-safe-bitmasks/

--- a/engine/src/gfx/background.cpp
+++ b/engine/src/gfx/background.cpp
@@ -344,7 +344,7 @@ void Background::Draw()
                 if (tex == NULL)
                     tex = _Universe->getLightMap();
                 const int    numpasses = 1;
-                static const float edge_fixup =
+                static float edge_fixup =
                     XMLSupport::parse_float( vs_config->getVariable( "graphics", "background_edge_fixup", "0" ) );
                 const float  ms      = 0.f, Ms = 1.f-edge_fixup/tex->boundSizeX;
                 const float  mt      = 0.f, Mt = 1.f-edge_fixup/tex->boundSizeY;

--- a/engine/src/gfx/background.cpp
+++ b/engine/src/gfx/background.cpp
@@ -344,8 +344,8 @@ void Background::Draw()
                 if (tex == NULL)
                     tex = _Universe->getLightMap();
                 const int    numpasses = 1;
-                static float edge_fixup =
-                    XMLSupport::parse_float( vs_config->getVariable( "graphics", "background_edge_fixup", "0" ) );
+                static float edge_fixup = 0.f;
+                    // XMLSupport::parse_float( vs_config->getVariable( "graphics", "background_edge_fixup", "0" ) );
                 const float  ms      = 0.f, Ms = 1.f-edge_fixup/tex->boundSizeX;
                 const float  mt      = 0.f, Mt = 1.f-edge_fixup/tex->boundSizeY;
                 const float  _stca[] = {-1.f, -Ms, ms, Ms, +1.f}, _ttca[] = {-1.f, -Mt, mt, Mt, +1.f};

--- a/engine/src/gfx/background.cpp
+++ b/engine/src/gfx/background.cpp
@@ -344,8 +344,8 @@ void Background::Draw()
                 if (tex == NULL)
                     tex = _Universe->getLightMap();
                 const int    numpasses = 1;
-                static float edge_fixup = 0.f;
-                    // XMLSupport::parse_float( vs_config->getVariable( "graphics", "background_edge_fixup", "0" ) );
+                static float edge_fixup =
+                    XMLSupport::parse_float( vs_config->getVariable( "graphics", "background_edge_fixup", "0" ) );
                 const float  ms      = 0.f, Ms = 1.f-edge_fixup/tex->boundSizeX;
                 const float  mt      = 0.f, Mt = 1.f-edge_fixup/tex->boundSizeY;
                 const float  _stca[] = {-1.f, -Ms, ms, Ms, +1.f}, _ttca[] = {-1.f, -Mt, mt, Mt, +1.f};

--- a/engine/src/gfx/vdu.cpp
+++ b/engine/src/gfx/vdu.cpp
@@ -225,7 +225,7 @@ static void DrawHUDSprite( VDU *thus,
         s->SetSize( w, invertsprite ? -h : h );
         Texture *spritetex = s->getTexture();
         if (drawsprite && spritetex) {
-            static const float middle_point =
+            static float middle_point =
                 XMLSupport::parse_float( vs_config->getVariable( "graphics", "hud", "armor_hull_size", ".55" ) );
             static bool top_view = XMLSupport::parse_bool( vs_config->getVariable( "graphics", "hud", "top_view", "false" ) );
             float  middle_point_small = 1-middle_point;

--- a/engine/src/gfx/vdu.cpp
+++ b/engine/src/gfx/vdu.cpp
@@ -1615,7 +1615,7 @@ void VDU::DrawWeapon( Unit *parent )
     }
     if (mbuf.length() != mlen)
         buf += mbuf;
-    static float background_alpha
+    static float background_alpha = 
         XMLSupport::parse_float( vs_config->getVariable( "graphics", "hud", "text_background_alpha", "0.0625" ) );
     GFXColor     tpbg = tp->bgcol;
     bool automatte    = (0 == tpbg.a);

--- a/engine/src/gfx/vdu.cpp
+++ b/engine/src/gfx/vdu.cpp
@@ -1615,8 +1615,8 @@ void VDU::DrawWeapon( Unit *parent )
     }
     if (mbuf.length() != mlen)
         buf += mbuf;
-    static float background_alpha = 0.0625f;
-        // XMLSupport::parse_float( vs_config->getVariable( "graphics", "hud", "text_background_alpha", "0.0625" ) );
+    static float background_alpha
+        XMLSupport::parse_float( vs_config->getVariable( "graphics", "hud", "text_background_alpha", "0.0625" ) );
     GFXColor     tpbg = tp->bgcol;
     bool automatte    = (0 == tpbg.a);
     if (automatte) tp->bgcol = GFXColor( 0, 0, 0, background_alpha );

--- a/engine/src/gfx/vdu.cpp
+++ b/engine/src/gfx/vdu.cpp
@@ -1615,8 +1615,8 @@ void VDU::DrawWeapon( Unit *parent )
     }
     if (mbuf.length() != mlen)
         buf += mbuf;
-    static float background_alpha =
-        XMLSupport::parse_float( vs_config->getVariable( "graphics", "hud", "text_background_alpha", "0.0625" ) );
+    static float background_alpha = 0.0625f;
+        // XMLSupport::parse_float( vs_config->getVariable( "graphics", "hud", "text_background_alpha", "0.0625" ) );
     GFXColor     tpbg = tp->bgcol;
     bool automatte    = (0 == tpbg.a);
     if (automatte) tp->bgcol = GFXColor( 0, 0, 0, background_alpha );

--- a/engine/src/gfxlib_struct.h
+++ b/engine/src/gfxlib_struct.h
@@ -626,7 +626,7 @@ public:
         attenuate[1] = attenuate[2] = 0;
         diffuse[0]   = diffuse[1] = diffuse[2] = 0; //openGL defaults
         specular[0]  = specular[1] = specular[2] = 0; //openGL defaults
-        ambient[0]   = ambient[1] = ambient[2] = options = 0;
+        ambient[0]   = ambient[1] = ambient[2] = 0;
         diffuse[3]   = specular[3] = ambient[3] = 1;
         target = -1; //physical GL light its saved in
 
@@ -635,6 +635,8 @@ public:
         cutoff = 180.0f;
         size = 0.0f;
         occlusion = 1.f;
+        options = 0;
+        apply_attenuate( attenuated() );
     }
 
     GFXLight( const bool enabled, const GFXColor &vect,
@@ -675,6 +677,7 @@ public:
         cutoff = other.cutoff;
         size = other.size;
         occlusion = other.occlusion;
+        options = other.options;
     };
 
     void SetProperties( enum LIGHT_TARGET, const GFXColor &color );

--- a/engine/src/gldrv/gl_light.cpp
+++ b/engine/src/gldrv/gl_light.cpp
@@ -109,10 +109,11 @@ GFXLight::GFXLight( const bool enabled,
     this->cutoff = cutoff;
     this->size   = size;
     apply_attenuate( attenuated() );
-    if (enabled)
+    if (enabled) {
         this->enable();
-    else
+    } else {
         this->disable();
+    }
 }
 
 void GFXLight::disable()
@@ -372,8 +373,9 @@ void /*GFXDRVAPI*/ GFXSetLightContext( const int con_number )
 void GFXDestroyAllLights()
 {
     lighttable.Clear();
-    if (GLLights)
+    if (GLLights) {
         free( GLLights );
+    }
 }
 
 static void SetupGLLightGlobals()
@@ -382,8 +384,9 @@ static void SetupGLLightGlobals()
     glGetIntegerv( GL_MAX_LIGHTS, &GFX_MAX_LIGHTS );
     if (!GLLights) {
         GLLights = (OpenGLLights*) malloc( sizeof (OpenGLLights)*GFX_MAX_LIGHTS );
-        for (int i = 0; i < GFX_MAX_LIGHTS; i++)
+        for (int i = 0; i < GFX_MAX_LIGHTS; i++) {
             GLLights[i].index = -1;
+        }
     }
 
     GFXSetCutoff( game_options.lightcutoff );

--- a/engine/src/gldrv/gl_light_state.cpp
+++ b/engine/src/gldrv/gl_light_state.cpp
@@ -238,8 +238,9 @@ bool gfx_light::Create( const GFXLight &temp, bool global )
 void gfx_light::Kill()
 {
     Disable();     //first disables it...which _will_ remove it from the light table.
-    if (target >= 0)
+    if (target >= 0) {
         TrashFromGLLights();          //then if not already done, trash from GLlights;
+    }
     target  = -2;
     options = 0;
 }
@@ -361,10 +362,12 @@ void gfx_light::ResetProperties( const enum LIGHT_TARGET light_targ, const GFXCo
         t.SetProperties( light_targ, color );
         changed = RemoveFromTable( false, t );
         *this = t;
-        if (changed)
+        if (changed) {
             AddToTable();
-        if (target >= 0)
+        }
+        if (target >= 0) {
             TrashFromGLLights();
+        }
         return;
     }
     switch (light_targ)

--- a/engine/src/gldrv/gl_light_state.cpp
+++ b/engine/src/gldrv/gl_light_state.cpp
@@ -156,8 +156,8 @@ Hashtable3d< LineCollideStar, 20, CTACC, lighthuge >lighttable;
 GFXLight gfx_light::operator=( const GFXLight &tmp )
 {   // Let's see if I can write a better copy operator
 //    memcpy( this, &tmp, sizeof (GFXLight) );
-            this->target  = -1;
-            this->options = 0;
+            this->target  = tmp.target;
+            this->options = tmp.options;
             memcpy( this->vect, tmp.vect, sizeof (float)*3 );
             memcpy( this->diffuse, tmp.diffuse, sizeof (float)*4 );
             memcpy( this->specular, tmp.specular, sizeof (float)*4 );
@@ -168,11 +168,12 @@ GFXLight gfx_light::operator=( const GFXLight &tmp )
             this->cutoff = tmp.cutoff;
             this->size   = tmp.size;
             this->occlusion = tmp.occlusion;
-            apply_attenuate( attenuated() );
-            if (GFX_LIGHT_ENABLED) //This is always true...No reason to define a light and not turn it on...
-                this->enable();
-            else
-                this->disable();
+            apply_attenuate( tmp.attenuated() );
+            // if (tmp.enabled()) {
+            //     this->enable();
+            // } else {
+            //     this->disable();
+            // }
             return tmp;
 }
 

--- a/engine/src/star_system.h
+++ b/engine/src/star_system.h
@@ -142,7 +142,7 @@ protected:
 public:
     // Constructors
     StarSystem( const string filename, const Vector &centroid = Vector( 0, 0, 0 ), const float timeofyear = 0 );
-    ~StarSystem();
+    virtual ~StarSystem();
     friend class Universe;
 
     // Methods

--- a/engine/src/vs_globals.h
+++ b/engine/src/vs_globals.h
@@ -97,7 +97,7 @@ public:
     }
     MyType operator[]( unsigned int i )
     {
-        // Enforce bounds checking
+        // stephengtuggy 2020-10-17: Enforce bounds checking
         return active_missions->at(i);
     }
     ~LeakVector()

--- a/engine/src/vs_globals.h
+++ b/engine/src/vs_globals.h
@@ -97,7 +97,8 @@ public:
     }
     MyType operator[]( unsigned int i )
     {
-        return (*active_missions)[i];
+        // Enforce bounds checking
+        return active_missions->at(i);
     }
     ~LeakVector()
     {

